File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.py
def solution(keymap, targets):
    answer = [0] * len(targets)
    key_dict = {}
    
    # key, 가장 빠른 순서로 dict 만듦
    for key in keymap:
        for k in key:
            key_dict[k] = min(key.find(k)+1, key_dict.get(k, len(key)))
    
    for idx, target in enumerate(targets):
        for t in target:
            if t in key_dict:
                answer[idx] += key_dict.get(t)
            else:
                answer[idx] = -1
                break
    return answer

File: 프로그래머스/1/160586. 대충 만든 자판/대충 만든 자판.java
import java.util.*;
class Solution {
    public int[] solution(String[] keymap, String[] targets) {
        int[] answer = new int[targets.length];
        Map<Character, Integer> map = new HashMap<>();
        
        for (String key : keymap) {
            for (int i = 0; i < key.length(); i++) {
                if (map.containsKey(key.charAt(i))) { // map에 해당 char이 있다면
                    if (map.get(key.charAt(i)) > i+1) { // 이미 저장된 값보다 i+1이 작다면
                        map.put(key.charAt(i), i+1); // i+1로 업데이트
                    }
                } else {
                    map.put(key.charAt(i), i+1);
                }
            }
        }
        
        for (int i = 0; i < targets.length; i++) {
            int sum = 0;
            for (int j = 0; j < targets[i].length(); j++) {
                if (!map.containsKey(targets[i].charAt(j))) { // map에 없는 알파벳이면
                    sum = -1;
                    break;
                } else {
                    sum += map.get(targets[i].charAt(j));
                }
            } answer[i] = sum;
        }
        return answer;
    }
}